File: ps/minimum-area-rectangle.cpp
struct hash_pair{
    size_t operator()(const pair<int, int> &pair) const{
        return hash<int>()(pair.first) ^ hash<int>()(pair.second);
    }
};
class Solution {
public:
    int minAreaRect(vector<vector<int>>& points) {
        unordered_map<pair<int, int>, int,hash_pair> map;
        for(auto point: points) {
            map[pair(point[0],point[1])] = 1;
        }
        int i,j;
        int res = 100000000;
        for(i=0;i<points.size();i++) {
            for(j=i+1;j<points.size();j++) {
                if (points[i][0] == points[j][0] || 
                   points[i][1] == points[j][1])
                    continue;
                if (map.find(pair(points[i][0],points[j][1])) == map.end() ||
                   map.find(pair(points[j][0],points[i][1])) == map.end()) 
                    continue;
                res = min(res, abs(points[j][0]-points[i][0])*abs(points[i][1]-points[j][1]));
                
            }
        }
        if (res == 100000000) return 0;
        return res;
    }
};

File: ps/minimum-area-rectangle.py
import sys


class Solution:
    def minAreaRect(self, points: List[List[int]]) -> int:
        # Time complexity 분석
        mp = {}
        for point in points:
            if point[0] not in mp:
                mp[point[0]] = {}
            mp[point[0]][point[1]] = 1
        min_sq = sys.maxsize

        passedx = {}
        for x in sorted(mp):
            ys = sorted(mp[x])
            for i, y1 in enumerate(ys):
                for j in range(i + 1, len(ys)):
                    y2 = ys[j]
                    if (y1, y2) in passedx:
                        min_sq = min(min_sq, abs((x - passedx[(y1, y2)]) * (y2 - y1)))
                    passedx[
                        (y1, y2)] = x  # store last x since the previous ones are no longer helpful for minimum square

        if min_sq == sys.maxsize:
            return 0
        return min_sq

    def minAreaRect2(self, points: List[List[int]]) -> int:
        mp = {}
        for point in points:
            if point[0] not in mp:
                mp[point[0]] = {}
            mp[point[0]][point[1]] = 1
        min_sq = sys.maxsize

        for i, p1 in enumerate(points):
            for j in range(i + 1, len(points)):
                p2 = points[j]
                if p1[0] == p2[0] or p1[1] == p2[1]:
                    continue

                if p2[1] not in mp[p1[0]] or p1[1] not in mp[p2[0]]:
                    continue

                min_sq = min(min_sq, abs((p1[0] - p2[0]) * (p1[1] - p2[1])))

        if min_sq == sys.maxsize:
            return 0
        return min_sq